core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1448,18 +1401,97 @@ private void handleErrorInSyncContext(Status error) {
   }
 
   private final class SubchannelImpl extends AbstractSubchannel {
-    // Set right after SubchannelImpl is created.
+    final CreateSubchannelArgs args;
+    final LbHelperImpl helper;
+    SubchannelStateListener listener;
     InternalSubchannel subchannel;
-    final Object shutdownLock = new Object();
-    final Attributes attrs;
+    boolean started;
+    boolean shutdown;
+    ScheduledHandle delayedShutdownTask;
+
+    SubchannelImpl(CreateSubchannelArgs args, LbHelperImpl helper) {
+      this.args = checkNotNull(args, "args");
+      this.helper = checkNotNull(helper, "helper");
+    }
 
-    @GuardedBy("shutdownLock")
-    boolean shutdownRequested;
-    @GuardedBy("shutdownLock")
-    ScheduledFuture<?> delayedShutdownTask;
+    @Override
+    public void start(final SubchannelStateListener listener) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      checkState(!started, "already started");
+      checkState(!shutdown, "already shutdown");
+      started = true;
+      this.listener = listener;
+      if (terminating) {
+        syncContext.execute(new Runnable() {
+            @Override
+            public void run() {
+              listener.onSubchannelState(ConnectivityStateInfo.forNonError(SHUTDOWN));
```

Comment:
The javadoc says `SHUTDOWN can be safely ignored`, do we need call `onSubchannelState(SHUTDOWN)`?

Reply:
The paragraph where this statement is from is the continuation of the previous paragraph, which talks about the need to create a picker for each update. Although a LoadBalancer usually don't need to react to a SHUTDOWN update, the ChannelImpl still need to propagate it for completeness. And in some cases e.g., HealthCheckingLoadBalancer and CachedSubchannelHelper, the SHUTDOWN update is actually consumed for cleanups. I have rephrased the javadoc to prevent confusion.